File: backend/app/services/llm/openai_provider.py
from typing import AsyncIterator, Optional

from openai import AsyncOpenAI

from app.core.config import settings
from app.core.exceptions import LLMException
from app.services.llm.base import BaseLLMProvider, LLMResponse, StreamChunk


class OpenAIProvider(BaseLLMProvider):
    """OpenAI LLM 提供商"""

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        model: Optional[str] = None,
    ):
        self.api_key = api_key or settings.openai_api_key
        self.base_url = base_url or settings.openai_base_url or None
        self.model = model or settings.openai_model

        if not self.api_key:
            raise LLMException("OpenAI API key is not configured")

        self.client = AsyncOpenAI(api_key=self.api_key, base_url=self.base_url)

    async def chat(self, messages: list[dict], **kwargs) -> LLMResponse:
        """
        非流式聊天

        Args:
            messages: 消息列表
            **kwargs: 额外参数（temperature, max_tokens, system_prompt 等）

        Returns:
            LLMResponse
        """
        try:
            system_prompt = kwargs.pop("system_prompt", None)
            full_messages = self._build_messages(system_prompt, messages)

            response = await self.client.chat.completions.create(
                model=self.model,
                messages=full_messages,
                stream=False,
                **kwargs,
            )

            choice = response.choices[0]
            return LLMResponse(
                content=choice.message.content or "",
                finish_reason=choice.finish_reason or "stop",
                usage={
                    "prompt_tokens": response.usage.prompt_tokens if response.usage else 0,
                    "completion_tokens": response.usage.completion_tokens if response.usage else 0,
                    "total_tokens": response.usage.total_tokens if response.usage else 0,
                },
            )
        except Exception as e:
            raise LLMException(f"OpenAI API error: {str(e)}")

    async def chat_stream(self, messages: list[dict], **kwargs) -> AsyncIterator[StreamChunk]:
        """
        流式聊天

        Args:
            messages: 消息列表
            **kwargs: 额外参数

        Yields:
            StreamChunk
        """
        try:
            system_prompt = kwargs.pop("system_prompt", None)
            full_messages = self._build_messages(system_prompt, messages)

            stream = await self.client.chat.completions.create(
                model=self.model,
                messages=full_messages,
                stream=True,
                **kwargs,
            )

            async for chunk in stream:
                if chunk.choices and len(chunk.choices) > 0:
                    delta = chunk.choices[0].delta
                    finish_reason = chunk.choices[0].finish_reason

                    content = delta.content or ""
                    if content or finish_reason:
                        yield StreamChunk(
                            content=content,
                            finish_reason=finish_reason,
                        )
        except Exception as e:
            raise LLMException(f"OpenAI streaming error: {str(e)}")

File: backend/src/modules/llm/providers/openai.provider.ts
import { Injectable, Logger } from '@nestjs/common';
import OpenAI from 'openai';
import { BaseLLMProvider, LLMResponse, StreamChunk } from './base.provider';
import { AppConfigService } from '../../../config/config.service';

@Injectable()
export class OpenAIProvider extends BaseLLMProvider {
  private readonly client: OpenAI;
  private readonly model: string;
  private readonly logger = new Logger(OpenAIProvider.name);

  constructor(private configService: AppConfigService) {
    super();
    const apiKey = configService.openaiApiKey;
    const baseUrl = configService.openaiBaseUrl || undefined;
    this.model = configService.openaiModel;

    this.client = new OpenAI({
      apiKey,
      baseURL: baseUrl,
    });
  }

  private buildMessages(systemPrompt: string | undefined, messages: any[]): OpenAI.ChatCompletionMessageParam[] {
    const result: OpenAI.ChatCompletionMessageParam[] = [];
    if (systemPrompt) {
      result.push({ role: 'system', content: systemPrompt });
    }
    for (const msg of messages) {
      result.push({
        role: msg.role as any,
        content: msg.content,
      });
    }
    return result;
  }

  async chat(messages: any[], systemPrompt?: string): Promise<LLMResponse> {
    const fullMessages = this.buildMessages(systemPrompt, messages);

    try {
      const response = await this.client.chat.completions.create({
        model: this.model,
        messages: fullMessages,
        stream: false,
      });

      const choice = response.choices[0];
      return {
        content: choice?.message?.content || '',
        finish_reason: choice?.finish_reason || 'stop',
        usage: response.usage
          ? {
              prompt_tokens: response.usage.prompt_tokens,
              completion_tokens: response.usage.completion_tokens,
              total_tokens: response.usage.total_tokens,
            }
          : undefined,
      };
    } catch (error) {
      this.logger.error(`OpenAI chat error: ${error.message}`, error.stack);
      throw error;
    }
  }

  async *chatStream(messages: any[], systemPrompt?: string): AsyncGenerator<StreamChunk> {
    const fullMessages = this.buildMessages(systemPrompt, messages);

    try {
      const stream = await this.client.chat.completions.create({
        model: this.model,
        messages: fullMessages,
        stream: true,
      });

      for await (const chunk of stream) {
        const delta = chunk.choices[0]?.delta;
        const finishReason = chunk.choices[0]?.finish_reason;

        if (delta?.content) {
          yield {
            content: delta.content,
            finish_reason: finishReason || null,
          };
        }

        if (finishReason) {
          yield {
            content: '',
            finish_reason: finishReason,
          };
        }
      }
    } catch (error) {
      this.logger.error(`OpenAI stream error: ${error.message}`, error.stack);
      throw error;
    }
  }
}
